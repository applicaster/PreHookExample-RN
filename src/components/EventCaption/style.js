import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  eventCaption: {
    fontSize: 14,
    paddingHorizontal: 12,
    paddingTop: 17,
    paddingBottom: 5,
    lineHeight: 18,
  },
});
