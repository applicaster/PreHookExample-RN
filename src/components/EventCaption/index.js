import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Text } from 'react-native';
import reactStringReplace from 'react-string-replace';
import { sendAnalyticEvent } from 'react-native-zapp-bridge';
import {
  OPEN_WEBVIEW_FROM_EVENT_CAPTION,
} from '../../constants/analyticEvents';
import { styles } from './style';

class EventCaption extends Component {
  constructor(props) {
    super(props);
    this.navigateToWebview = this.navigateToWebview.bind(this);
  }

  processCaption() {
    let { caption } = this.props;
    caption = this.highlightHashtagsAndUsers(caption);
    caption = this.highlightLinks(caption);
    return caption;
  }

  highlightHashtagsAndUsers(caption) {
    const highlightColor = { color: this.context.mainColor };
    const regex = /([@|#][\w](?:(?:[\w]|(?:\.(?!\.))){0,28}(?:[\w]))?)/g;
    const matchCallback = (match, i) => (<Text key={i} style={highlightColor}>{match}</Text>);
    return reactStringReplace(caption, regex, matchCallback);
  }

  highlightLinks(caption) {
    const linkStyle = { color: this.context.mainColor, textDecorationLine: 'underline' };
    const regex = /([(https|http)]*:\/\/[-A-Z0-9+&@#/%?=~_|!:,.;]*[-A-Z0-9+&@#/%=~_|])/ig;
    const matchCallback = (match, i) =>
      (<Text onPress={() => this.navigateToWebview(match)} key={`${i}-link`} style={linkStyle}>{match}</Text>);

    return reactStringReplace(caption, regex, matchCallback);
  }

  navigateToWebview(url) {
    const { navigation } = this.context;
    sendAnalyticEvent(OPEN_WEBVIEW_FROM_EVENT_CAPTION, { url }).then().catch();
    navigation.navigate('GenericWebView', { headerTitle: 'web', url });
  }
  
  render() {
    const { caption } = this.props;
    const captionColor = { color: this.context.textColor };
    return (caption)
    ? <Text style={[styles.eventCaption, captionColor]}>
      {this.processCaption(caption)}
    </Text>
    : <Text />;
  }
}

EventCaption.propTypes = {
  caption: PropTypes.string,
};

EventCaption.contextTypes = {
  mainColor: PropTypes.string,
  navigation: PropTypes.object,
  textColor: PropTypes.string,
};

export default EventCaption;
