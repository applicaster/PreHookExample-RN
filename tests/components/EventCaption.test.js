import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import EventCaption from '../../src/components/EventCaption';

jest.mock('react-native-zapp-bridge', () => 'NativeZappBridge');

describe('EventCaption', () => {
  test('component renders correctly', () => {
    const html = renderer.create(
      <EventCaption caption={'hola'} />
    ).toJSON();
    expect(html).toMatchSnapshot();
  });

  test('component renders correctly with no caption', () => {
    const html = renderer.create(
      <EventCaption caption={null} />
    ).toJSON();
    expect(html).toMatchSnapshot();
  });

  test('component renders correctly with complex caption with users and hashtags', () => {
    const html = renderer.create(
      <EventCaption caption={'hola @cpinto7 #wowow #test_ @applicaster more stuff yeiah'} />
    ).toJSON();
    expect(html).toMatchSnapshot();
  });
});
