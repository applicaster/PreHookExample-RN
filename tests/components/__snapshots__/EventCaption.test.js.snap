// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventCaption component renders correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#FFFFFF",
      "fontSize": 14,
      "lineHeight": 18,
      "paddingBottom": 12,
      "paddingHorizontal": 12,
      "paddingTop": 17,
    }
  }
>
  hola
</Text>
`;

exports[`EventCaption component renders correctly with complex caption with users and hashtags 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#FFFFFF",
      "fontSize": 14,
      "lineHeight": 18,
      "paddingBottom": 12,
      "paddingHorizontal": 12,
      "paddingTop": 17,
    }
  }
>
  hola 
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#7ED321",
      }
    }
  >
    @cpinto7
  </Text>
   
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#7ED321",
      }
    }
  >
    #wowow
  </Text>
   
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#7ED321",
      }
    }
  >
    #test_
  </Text>
   
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#7ED321",
      }
    }
  >
    @applicaster
  </Text>
   more stuff yeiah
</Text>
`;

exports[`EventCaption component renders correctly with no caption 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
/>
`;

exports[`EventCaption component renders correctly with no caption text 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
/>
`;
